<a href="https://colab.research.google.com/github/Spyke2022/Spyke2022/blob/main/Ativ_2_Prev_GRAF_SILAS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
rodovias = pd.read_csv('/content/rotas.csv')

In [ ]:
rodovias

DIRIGIBILIDADE   ROTAS   DISTANCIA      FALHAS    LOMBADAS  ILUMINACAO  \
0        continuous  string  continuous  continuous  continuous  continuous   
1             class    meta         NaN         NaN         NaN         NaN   
2               3.0      R1       159.0        20.0        16.0         4.0   
3               2.0      R2       131.0         3.0         2.0         1.0   
4               3.0      R3       141.0         2.0        27.0         3.0   
...             ...     ...         ...         ...         ...         ...   
5116            1.0   R5115       342.0         3.0        17.0         4.0   
5117            1.0   R5116       249.0        17.0        10.0         5.0   
5118            3.0   R5117        89.0        18.0        20.0         3.0   
5119            3.0   R5118       197.0         3.0        19.0         4.0   
5120            1.0   R5119       126.0         8.0        23.0         4.0   

     SINALIZACAO PEDAGIOS POSTOPOLICIAL  
0     continuous  0.0 1.0      discrete  
1            NaN      NaN           NaN  
2            3.0      1.0           0.0  
3            6.0      1.0           0.0  
4           57.0      1.0           0.0  
...          ...      ...           ...  
5116        24.0      0.0           0.0  
5117        25.0      0.0           0.0  
5118        56.0      1.0           0.0  
5119        71.0      0.0           0.0  
5120        12.0      1.0           0.0  

[5121 rows x 9 columns]

In [ ]:
#tipo de cada coluna do dataframe:
rodovias.iloc[0:2]

DIRIGIBILIDADE   ROTAS   DISTANCIA      FALHAS    LOMBADAS  ILUMINACAO  \
0     continuous  string  continuous  continuous  continuous  continuous   
1          class    meta         NaN         NaN         NaN         NaN   

  SINALIZACAO PEDAGIOS POSTOPOLICIAL  
0  continuous  0.0 1.0      discrete  
1         NaN      NaN           NaN

In [ ]:
#tipo de cada coluna do dataframe:
rodovias = rodovias.iloc[2:]

In [ ]:
rodovias

DIRIGIBILIDADE  ROTAS DISTANCIA FALHAS LOMBADAS ILUMINACAO SINALIZACAO  \
2               3.0     R1     159.0   20.0     16.0        4.0         3.0   
3               2.0     R2     131.0    3.0      2.0        1.0         6.0   
4               3.0     R3     141.0    2.0     27.0        3.0        57.0   
5               3.0     R4     269.0   13.0     20.0        1.0        93.0   
6               2.0     R5     182.0   27.0     10.0        4.0        78.0   
...             ...    ...       ...    ...      ...        ...         ...   
5116            1.0  R5115     342.0    3.0     17.0        4.0        24.0   
5117            1.0  R5116     249.0   17.0     10.0        5.0        25.0   
5118            3.0  R5117      89.0   18.0     20.0        3.0        56.0   
5119            3.0  R5118     197.0    3.0     19.0        4.0        71.0   
5120            1.0  R5119     126.0    8.0     23.0        4.0        12.0   

     PEDAGIOS POSTOPOLICIAL  
2         1.0           0.0  
3         1.0           0.0  
4         1.0           0.0  
5         0.0           0.0  
6         1.0           0.0  
...       ...           ...  
5116      0.0           0.0  
5117      0.0           0.0  
5118      1.0           0.0  
5119      0.0           0.0  
5120      1.0           0.0  

[5119 rows x 9 columns]

In [ ]:
print((type)(rodovias['FALHAS']), type(['ROTAS']))

<class 'pandas.core.series.Series'> <class 'list'>


In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv('/content/rotas.csv')

In [ ]:
df_1 = pd.crosstab(index=df['FALHAS'], columns=df['ROTAS'])

In [ ]:
df_1

ROTAS       R1  R10  R100  R1000  R1001  R1002  R1003  R1004  R1005  R1006  \
FALHAS                                                                       
0.0          0    1     0      0      0      0      0      0      0      0   
1.0          0    0     0      0      0      0      0      0      0      0   
10.0         0    0     0      0      0      0      0      0      0      0   
11.0         0    0     0      0      0      0      0      0      0      0   
12.0         0    0     0      0      0      0      0      0      0      1   
13.0         0    0     0      0      0      0      0      1      0      0   
14.0         0    0     0      0      0      0      0      0      0      0   
15.0         0    0     0      0      0      1      0      0      0      0   
16.0         0    0     0      0      0      0      0      0      0      0   
17.0         0    0     0      0      0      0      0      0      0      0   
18.0         0    0     0      0      0      0      0      0      0      0   
19.0         0    0     0      0      0      0      0      0      0      0   
2.0          0    0     0      0      0      0      0      0      0      0   
20.0         1    0     0      0      0      0      0      0      0      0   
21.0         0    0     0      0      0      0      0      0      0      0   
22.0         0    0     0      0      0      0      0      0      0      0   
23.0         0    0     0      0      0      0      0      0      0      0   
24.0         0    0     0      0      0      0      0      0      1      0   
25.0         0    0     0      0      0      0      0      0      0      0   
26.0         0    0     0      0      0      0      1      0      0      0   
27.0         0    0     0      0      0      0      0      0      0      0   
28.0         0    0     0      0      0      0      0      0      0      0   
29.0         0    0     0      0      0      0      0      0      0      0   
3.0          0    0     1      0      0      0      0      0      0      0   
30.0         0    0     0      0      0      0      0      0      0      0   
4.0          0    0     0      0      0      0      0      0      0      0   
5.0          0    0     0      0      1      0      0      0      0      0   
6.0          0    0     0      1      0      0      0      0      0      0   
7.0          0    0     0      0      0      0      0      0      0      0   
8.0          0    0     0      0      0      0      0      0      0      0   
9.0          0    0     0      0      0      0      0      0      0      0   
continuous   0    0     0      0      0      0      0      0      0      0   

ROTAS       ...  R991  R992  R993  R994  R995  R996  R997  R998  R999  string  
FALHAS      ...                                                                
0.0         ...     0     0     0     0     0     0     0     0     0       0  
1.0         ...     0     0     0     0     0     0     0     0     0       0  
10.0        ...     0     0     0     0     0     0     0     0     0       0  
11.0        ...     0     0     0     0     0     0     0     0     0       0  
12.0        ...     0     0     0     0     0     0     0     0     0       0  
13.0        ...     0     0     0     0     0     0     0     0     0       0  
14.0        ...     0     0     0     0     0     0     0     0     0       0  
15.0        ...     0     0     0     0     0     0     0     0     0       0  
16.0        ...     0     0     0     0     0     0     0     0     0       0  
17.0        ...     0     0     0     0     0     0     0     1     0       0  
18.0        ...     0     0     1     0     0     0     0     0     0       0  
19.0        ...     0     0     0     0     1     0     0     0     0       0  
2.0         ...     0     0     0     0     0     0     0     0     0       0  
20.0        ...     0     0     0     0     0     0     0     0     0       0  
21.0        ...     0     0     0     0     0     0     0     0    

In [ ]:
# prompt: Usando o DataFrame df_1: Novo DataFrame com ordem decrescente de falhas

df_2 = df_1.sort_values(by='FALHAS', ascending=False)
df_2


ROTAS       R1  R10  R100  R1000  R1001  R1002  R1003  R1004  R1005  R1006  \
FALHAS                                                                       
continuous   0    0     0      0      0      0      0      0      0      0   
9.0          0    0     0      0      0      0      0      0      0      0   
8.0          0    0     0      0      0      0      0      0      0      0   
7.0          0    0     0      0      0      0      0      0      0      0   
6.0          0    0     0      1      0      0      0      0      0      0   
5.0          0    0     0      0      1      0      0      0      0      0   
4.0          0    0     0      0      0      0      0      0      0      0   
30.0         0    0     0      0      0      0      0      0      0      0   
3.0          0    0     1      0      0      0      0      0      0      0   
29.0         0    0     0      0      0      0      0      0      0      0   
28.0         0    0     0      0      0      0      0      0      0      0   
27.0         0    0     0      0      0      0      0      0      0      0   
26.0         0    0     0      0      0      0      1      0      0      0   
25.0         0    0     0      0      0      0      0      0      0      0   
24.0         0    0     0      0      0      0      0      0      1      0   
23.0         0    0     0      0      0      0      0      0      0      0   
22.0         0    0     0      0      0      0      0      0      0      0   
21.0         0    0     0      0      0      0      0      0      0      0   
20.0         1    0     0      0      0      0      0      0      0      0   
2.0          0    0     0      0      0      0      0      0      0      0   
19.0         0    0     0      0      0      0      0      0      0      0   
18.0         0    0     0      0      0      0      0      0      0      0   
17.0         0    0     0      0      0      0      0      0      0      0   
16.0         0    0     0      0      0      0      0      0      0      0   
15.0         0    0     0      0      0      1      0      0      0      0   
14.0         0    0     0      0      0      0      0      0      0      0   
13.0         0    0     0      0      0      0      0      1      0      0   
12.0         0    0     0      0      0      0      0      0      0      1   
11.0         0    0     0      0      0      0      0      0      0      0   
10.0         0    0     0      0      0      0      0      0      0      0   
1.0          0    0     0      0      0      0      0      0      0      0   
0.0          0    1     0      0      0      0      0      0      0      0   

ROTAS       ...  R991  R992  R993  R994  R995  R996  R997  R998  R999  string  
FALHAS      ...                                                                
continuous  ...     0     0     0     0     0     0     0     0     0       1  
9.0         ...     1     0     0     1     0     0     0     0     0       0  
8.0         ...     0     0     0     0     0     0     0     0     0       0  
7.0         ...     0     0     0     0     0     1     0     0     0       0  
6.0         ...     0     0     0     0     0     0     0     0     0       0  
5.0         ...     0     0     0     0     0     0     1     0     0       0  
4.0         ...     0     0     0     0     0     0     0     0     0       0  
30.0        ...     0     0     0     0     0     0     0     0     0       0  
3.0         ...     0     0     0     0     0     0     0     0     0       0  
29.0        ...     0     0     0     0     0     0     0     0     0       0  
28.0        ...     0     1     0     0     0     0     0     0     0       0  
27.0        ...     0     0     0     0     0     0     0     0     0       0  
26.0        ...     0     0     0     0     0     0     0     0     0       0  
25.0        ...     0     0     0     0     0     0     0     0     0       0  
24.0        ...     0     0     0     0     0     0     0     0    

In [ ]:
from builtins import sorted

**Verificando o número de linhas e de colunas do novo DataFrame df_1 filtrado com as colunas FALHAS X ROTAS**

In [ ]:
print(len(df_1.columns))
df_1.shape[0]

5120


32

*COLUNAS = 5120 ROTAS DIFERENTES
*FALHAS  = 32   QTD DE FALHAS DIFERENTES (INTERVALO QUE VAI DE 0 A 31 )

Como você tem uma lista de strings e quer ordená-las como inteiros, basta passar a classe int através do argumento key:

>>> sorted(numeros, key=int)
['1', '2', '3', '4', '10', '20', '30']
Assim, a lista será ordenada conforme o retorno de int para cada valor, mas o resultado permanecerá como uma lista de strings. Caso queira ordená-los de form,a reversa, basta informar o argumento reverse como verdadeiro:

>>> sorted(numeros, key=int, reverse=True)
['30', '20', '10', '4', '3', '2', '1']

In [ ]:
df_1['FALHAS'] = pd.to_numeric(df_1['FALHAS'], errors='coerce')

KeyError: 'FALHAS'

In [ ]:
!pip install pandas
import pandas as pd

# Check if the DataFrame 'df_1' exists
if 'df_1' not in globals():
    print("DataFrame 'df_1' is not defined.")
else:
    # Proceed with sorting the DataFrame
    df_1.sort_values(by='FALHAS', key=int, ascending=False)

TypeError: cannot convert the series to <class 'int'>

In [ ]:
df_1.sort_values(by='FALHAS', key=int, ascending=False)

In [ ]:
import pandas as pd

# Dados de exemplo
data = [23, 24, 22, 30, 29, 21, 20, 35, 40, 42,
        21, 30, 33, 21, 25, 25, 40, 39, 19, 31]

# Cria um DataFrame
df = pd.DataFrame(data, columns=['Idade'])

# Define os intervalos de classe
bins = [18, 25, 32, 39, 46]

# Cria a coluna de intervalos de classe
df['Intervalo de Classe'] = pd.cut(df['Idade'], bins)

# Cria a tabela de distribuição de frequências
tabela_freq = df['Intervalo de Classe'].value_counts(sort=False)

# Organiza os intervalos de classe em ordem crescente
tabela_freq = tabela_freq.sort_index()

# Calcula as frequências relativas
tabela_freq_relativa = tabela_freq / tabela_freq.sum()

# Calcula as frequências acumuladas
tabela_freq_acumulada = tabela_freq.cumsum()

# Combina todas as informações em uma tabela
tabela_final = pd.DataFrame({
    'Frequência': tabela_freq,
    'Frequência Relativa': tabela_freq_relativa,
    'Frequência Acumulada': tabela_freq_acumulada
})

print(tabela_final)

In [ ]:
import pandas as pd
# Dados de exemplo
data = [0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0, 19.0, 20.0, 21.0, 22.0, 23.0, 24.0, 25.0, 26.0, 27.0, 28.0, 29.0, 30.0, 31.0]
# Cria um DataFrame
df = pd.DataFrame(data, columns=['FALHAS'])

# Define os intervalos de classe
bins = [ 0, 5, 10, 15, 20, 25, 31 ]

# Cria a coluna de intervalos de classe
df['Intervalo de Classe'] = pd.cut(df['FALHAS'], bins)

# Cria a tabela de distribuição de frequências
tabela_freq = df['Intervalo de Classe'].value_counts(sort=False)

# Organiza os intervalos de classe em ordem crescente
tabela_freq = tabela_freq.sort_index()

# Calcula as frequências relativas
tabela_freq_relativa = tabela_freq / tabela_freq.sum()

# Calcula as frequências acumuladas
tabela_freq_acumulada = tabela_freq.cumsum()

# Combina todas as informações em uma tabela
tabela_final = pd.DataFrame({
    'Frequência': tabela_freq,
    'Frequência Relativa': tabela_freq_relativa,
    'Frequência Acumulada': tabela_freq_acumulada
})

print(tabela_final)

In [ ]:
if 'ROTAS' in df_1.columns:
    frequencia = df_1['ROTAS'].value_counts()
    percentual = df_1['ROTAS'].value_counts(normalize=True) * 100
else:
    print("Column 'FALHAS' not found in the DataFrame.")

In [ ]:
df_1.columns

In [ ]:
frequencia = df_1['FALHAS'].value_counts()
percentual = df_1['FALHAS'].value_counts(normalize = True)*100

In [ ]:
dist_freq_qualitativas = pd.DataFrame({'Frequência': frequencia, 'Porcentagem(%)': percentual})

In [ ]:
dist_freq_qualitativas.rename_axis('Sexo', axis = 'columns', inplace = True)

dist_freq_qualitativas

In [ ]:
from matplotlib import pyplot as plt
import seaborn as sns
df_0 = pd.read_csv("/content/rotas.csv")
df_0.groupby('FALHAS').size().plot(kind='barh', color=sns.palettes.mpl_palette('Dark2'))
plt.gca().spines[['top', 'right',]].set_visible(False)

In [ ]:
from matplotlib import pyplot as plt
import seaborn as sns
df_1 = pd.read_csv("/content/rotas.csv")
df_1.groupby('ROTAS').size().plot(kind='bar', color=sns.palettes.mpl_palette('Dark2'))
plt.gca().spines[['top', 'right',]].set_visible(False)

In [ ]:
from matplotlib import pyplot as plt
import seaborn as sns
import pandas as pd
plt.subplots(figsize=(14, 14))
df_2 = pd.read_csv("/content/rotas.csv")
df_3 = pd.read_csv("/content/rotas.csv")
df_2.groupby('ROTAS').size().plot(kind='bar', color=sns.palettes.mpl_palette('Dark2'))
df_2dhist = pd.DataFrame({
    x_label: grp['ROTAS'].value_counts()
    for x_label, grp in df_3.groupby('FALHAS')
})
sns.heatmap(df_2dhist, cmap='viridis')
plt.xlabel('ROTAS')
_ = plt.ylabel('FALHAS')

In [ ]:
rotas = rodovias['ROTAS']

In [ ]:
print(rotas[:5])

In [ ]:
rotas.head(5119).str.split('R')

In [ ]:
rotas = globals()["rotas"]

In [ ]:
import builtins

In [ ]:
value = rotas.iloc[0]

In [ ]:
print(value)

In [ ]:
try:
    int(value, base=10)
except ValueError:
    print("The value variable does not contain a valid integer representation in base 10.")

In [ ]:
value = value.replace(',', '')

In [ ]:
try:
    rotas2 = int(value, base=10)
except ValueError:
    print("The value variable does not contain a valid integer representation in base 10. Setting rotas2 to 0.")
    rotas2 = 0

In [ ]:
print(rotas2)

In [ ]:
rodovias = rodovias[rodovias["ROTAS"] == "string"]
rodovias["ROTAS"] = pd.to_numeric(rodovias["ROTAS"])

In [ ]:
rodovias["ROTAS"] = rodovias["ROTAS"].replace("string", 0)

In [ ]:
import pandas as pd
rodovias = pd.read_csv('rotas.csv')
colunas_numericas = ['DIRIGIBILIDADE', 'ROTAS','DISTANCIA', 'FALHAS', 'LOMBADAS', 'ILUMINACAO', 'SINALIZACAO']
rodovias_colunas_continuas = pd.DataFrame(df[colunas_numericas].apply(pd.to_numeric, errors='coerce'))

In [ ]:
rodovias.columns

In [ ]:
rodovias['ROTAS'].isna().sum()
rodovias['FALHAS'].isna().sum()

In [ ]:
rodovias['ROTAS'].dtype
rodovias['FALHAS'].dtype

In [ ]:
frequencia = rodovias_colunas_continuas[['FALHAS','ROTAS']].value_counts()
frequencia

In [ ]:
frequencia = rodovias_colunas_continuas[['FALHAS','DIRIGIBILIDADE']].value_counts()
frequencia


In [ ]:
rodovias.POSTOPOLICIAL.value_counts()

In [ ]:
#colunas continuas
rodovias_colunas_continuas = rodovias[['ROTAS','DIRIGIBILIDADE', 'DISTANCIA', 'FALHAS','ILUMINACAO']]

In [ ]:
rodovias_colunas_continuas.ROTAS.iloc[1]

In [ ]:
rodovias_colunas_continuas.info()

In [ ]:
rodovias_colunas_continuas

In [ ]:
df_colunas_continuas.corr()

In [ ]:
# análise de correlação
correlation = df_colunas_continuas.corr()

In [ ]:
import seaborn as sns

# plot da matriz de correlação
plot = sns.heatmap(correlation, annot = True, fmt=".3f", linewidths=.6)
plot

In [ ]:
import pandas as pd
import plotly.offline as py
import plotly.graph_objs as go
py.init_notebook_mode(connected=True)

In [ ]:
# Import the chart_studio.plotly module
from chart_studio.plotly import iplot

In [ ]:
rodovias = globals()['rodovias']

In [ ]:
print(rodovias.columns.tolist())

In [ ]:
print(df.columns.tolist())

In [ ]:
trace = go.Scatter(x = rodovias['ROTAS'],
                   y = rodovias['FALHAS'],
                   mode = 'markers',
                   marker = {'color' : '#27ae60',
                              'line' : {'width': 1,
                                        'color': '#8e44ad'}},
                   opacity=.8)
data = [trace]
py.iplot(data)

In [ ]:
# Gráfico usando apenas marcadores
trace1 = go.Scatter(x = ['DIRIGIBILIDADE', 'ROTAS', 'DISTANCIA', 'FALHAS', 'LOMBADAS', 'ILUMINACAO', 'SINALIZACAO', 'PEDAGIOS', 'POSTOPOLICIAL'],
                    y = [1, 2, 3, 4, 5],
                    mode = 'markers',
                    name = 'Apenas marcadores')
# Gráfico de apenas linhas
trace2 = go.Scatter(x = ['DIRIGIBILIDADE', 'ROTAS', 'DISTANCIA', 'FALHAS', 'LOMBADAS', 'ILUMINACAO', 'SINALIZACAO', 'PEDAGIOS', 'POSTOPOLICIAL'],
                    y = [11, 12, 13, 14, 15],
                    mode = 'lines',
                    name = 'Apenas linhas')
# Criando gráfico com marcadores e linhas
trace3 = go.Scatter(x = ['DIRIGIBILIDADE', 'ROTAS', 'DISTANCIA', 'FALHAS', 'LOMBADAS', 'ILUMINACAO', 'SINALIZACAO', 'PEDAGIOS', 'POSTOPOLICIAL'],
                    y = [6, 7, 8, 9, 10],
                    mode = 'markers+lines',
                    name = 'Marcadores e Linhas')
data = [trace1, trace2, trace3]
py.iplot(data)

O coeficiente de Pearson, também chamado de “coeficiente de correlação produto-momento” ou chamado de “ρ de Pearson”, mede o grau de correlação através do cálculo de direção positiva ou negativa. Este coeficiente, normalmente representado por ρ assume apenas valores entre -1 e 1.

A análise de correlação vai retornar três possíveis cenários: (1) correlação positiva; (2) correlação negativa; e (3) não há correlação.

Correlação positiva: quando duas variáveis que possuem correlação crescem ou decrescem juntas, ou seja, que possuem uma relação direta;
Correlação negativa: quando duas variáveis que possuem correlação mas quando uma variável cresce a outra decresce, ou vice-versa;
Não ter correlação: quando o crescimento ou decrescimento de uma variável não tem efeito sobre outra variável.


Como podemos interpretar os valores que ρ pode assumir?

ρ = 0,9 a 1 (positivo ou negativo): correlação muito forte;
ρ = 0,7 a 09 (positivo ou negativo): correlação forte;
ρ = 0,5 a 0,7 (positivo ou negativo): correlação moderada;
ρ = 0,3 a 0,5 (positivo ou negativo): correlação fraca;
ρ = 0 a 0,3 (positivo ou negativo): não possui correlação.


# **DIRIGIBILIDADE X LOMBADAS**

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

# Suponha que 'df' seja o seu DataFrame
sns.barplot(x='DIRIGIBILIDADE', y='LOMBADAS', palette='Set2', data=df_colunas_continuas[['DIRIGIBILIDADE', 'LOMBADAS']])
plt.show()

# **DIRIGIBILIDADE X DISTANCIA**

In [ ]:
import matplotlib.pyplot as plt
sns.barplot(x='DIRIGIBILIDADE', y='DISTANCIA', palette='Set3', data=df_colunas_continuas[['DIRIGIBILIDADE', 'DISTANCIA']])
plt.show()

# **DIRIGIBILIDADE X FALHAS**

In [ ]:
sns.barplot(x='DIRIGIBILIDADE', y='FALHAS', palette='Set1', data=df_colunas_continuas[['DIRIGIBILIDADE', 'FALHAS']])
plt.show()

# **DIRIGIBILIDADE X ILUMINACAO**

In [ ]:
sns.barplot(x='DIRIGIBILIDADE', y='ILUMINACAO', palette='Set2', data=df_colunas_continuas[['DIRIGIBILIDADE', 'ILUMINACAO']])
plt.show()

# **DIRIGIBILIDADE X SINALIZACAO**

In [ ]:
sns.barplot(x='DIRIGIBILIDADE', y='SINALIZACAO', palette='Set1', data=df_colunas_continuas[['DIRIGIBILIDADE', 'SINALIZACAO']])
plt.show()

# ** regressão multivariada é uma coleção de técnicas estatísticas que constroem modelos que descrevam de maneira razoável relações entre várias variáveis explicativas de um determinado processo. A diferença entre a regressão linear simples e a multivariada é que nesta são tratadas duas ou mais variáveis explicativas.**
**A modelagem de equações estruturais, denominada SEM (do inglês Structural Equation Modeling), é um método de análise multivariada que combina aspectos de regressão múltipla e de análise fatorial, com o objetivo de estimar simultaneamente uma série de relações de dependência.**